doc/user-manual/language/sort-system.lagda.rst

```diff
@@ -0,0 +1,150 @@
+..
+  ::
+  module sort.system where
+
+.. _sort-system:
+
+***********
+Sort System
+***********
+
+.. _intro-sorts:
+
+Agda's sorts system
+-------------------
+
+The implementation of Agda’s sorts system is closely based on the theory
+of pure type systems. Sorts (aka universes) are types whose members
+themselves are again types. The fundamental sort in Agda is named ``Set``
+and it denotes the universe of small types.
+
+But for some applications, other sorts are needed. Martin-Löf’s type theory
+had originally a rule ``Set : Set`` but Girard showed that it is inconsistent.
+This result is known as Girard’s paradox. Agda implements an
+:ref:`infinite hierarchy of universes <universe-levels>` that avoids this
+inconsistency.
+
+Under universe polymorphism, levels can be arbitrary terms, e.g., a
+level that contains free variables. Sometimes, we will have to check
+that some expression has a valid type without knowing what sort it has.
+
+funSort
+-------
+
+Agda’s internal representation of sorts implements a constructor (sort
+metavariable) representing an unknown sort. The constraint solver can
+compute these sort metavariables, just like it does when computing
+regular term metavariables.
+
+But this sort metavariable needs other constructors to solve function
+types. The constructor ``funSort`` computes the sort of a function type
```

Comment:
It should be made very clear that the sort constructors `funSort` `univSort` `piSort` are internal constructors that are sometimes printed, but can not be entered by the user.

Reply:
I added a note before the explanations.